techmatters/hrm
# Rework permissions files

Primary reviewer: @stephenhand.
This PR is easier to review commit by commit.

## Description
- This PR replicates the changes of https://github.com/techmatters/flex-plugins/pull/669 to be in sync. This way only `.json` files will describe permissions sets for helplines. For further information on the changes being introduced here, read the above linked PR.
- Moves all existing permissions files to the new format (copied over from the frontend definitions).
  
![Screenshot from 2022-03-14 22-06-41](https://user-images.githubusercontent.com/15805319/158286352-cbefb505-9aeb-4a17-b684-90a14385cbc8.png)

Notes:
- Backend and frontend are actually not fully in sync, since there is one action, "view post survey" that's supported by the backend but missing in the action types. I think that's fine for now. 
- [WIP] Another subtle difference is that `editChildIsAtRisk` and `editFollowUpDate` are present in the frontend, but are introduced to the backend in this PR. There is no actual support to decide if a case edit is changing one of this two properties.
- A followup PR will expose this definition files from an API. I think the best idea is to serve them from here, but TBD with the rest of the team.

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1146)
- [N/A] Strings are localized

### Verification steps
For this you'll need two counselors (or at least, a valid `token`for two different counselor). This is assuming you are using `zm` config, but can be any definition file for permissions rules.
- `npm run start` to start a local server.
- If you don't have any case in your local DB, you need to create one (easier from Flex, after exposing the local server with ngrok and changing `Twilio.Flex.Manager.getInstance().serviceConfiguration.attributes.hrm_base_url`).
- After the case is created, with the creator:
  - Try creating new "items" in the case from the Flex UI (add a note when there are no notes).
  - Try adding more items to an existing property (add a second note).
  - Grab the `token` and try editing some existing stuff (like the last name of a perpetrator). This can be done like
    ```curl -X PUT 'http://localhost:3000/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/cases/<Case ID>' -H "Content-Type: application/json" -H "Authorization: Bearer <the token>" -d '{ "info": { .... the case.info as it is in the DB plus the update  }'```
- Try doing all the same but with the token of a second counselor. If the configuration in use is `zm` all of them should fail (unauthorized). For the "add" steps you also need to curl (or deactivate the "hide buttons" behavior in the frontend). To do the first you can 
    ```curl -X PUT 'http://localhost:3000/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/cases/<Case ID>' -H "Content-Type: application/json" -H "Authorization: Bearer <the other token>" -d '{ "info": { .... the case.info as it is in the DB plus one more item in one the already existing properties (like a second note)  }'```
